vendor/brave-ios/Ledger/BATBraveLedger.mm

```diff
@@ -190,9 +193,129 @@ - (void)dealloc
   delete rewardsDatabase;
 }
 
+- (void)initializeLedgerService:(BOOL)executeMigrateScript
+{
+  if (self.initialized || self.initializing) {
+    return;
+  }
+  self.initializing = YES;
+  
+  BLOG(ledger::LogLevel::LOG_DEBUG) << "DB: Migrate from CoreData? " << (executeMigrateScript ? "YES" : "NO") << std::endl;
+  ledger->Initialize(executeMigrateScript, ^(ledger::Result result){
+    self.initialized = result == ledger::Result::LEDGER_OK;
+    self.initializing = NO;
+    if (self.initialized) {
+      self.prefs[kMigrationSucceeded] = @(YES);
+      [self savePrefs];
+    } else {
+      if (result == ledger::Result::DATABASE_INIT_FAILED) {
+        // Failed to migrate data...
+        switch (self.migrationType) {
+          case BATLedgerDatabaseMigrationTypeDefault:
+            BLOG(ledger::LogLevel::LOG_ERROR) << "DB: Full migration failed, attempting BAT only migration." << std::endl;
+            self.dataMigrationFailed = YES;
+            self.migrationType = BATLedgerDatabaseMigrationTypeTokensOnly;
+            [self resetRewardsDatabase];
+            // attempt re-initialize without other data
+            [self initializeLedgerService:YES];
+            return;
+          case BATLedgerDatabaseMigrationTypeTokensOnly:
+            BLOG(ledger::LogLevel::LOG_ERROR) << "DB: BAT only migration failed. Initializing without migration." << std::endl;
+            self.dataMigrationFailed = YES;
+            self.migrationType = BATLedgerDatabaseMigrationTypeNone;
+            [self resetRewardsDatabase];
+            // attempt initialize without migrating at all
+            [self initializeLedgerService:NO];
+            return;
+          default:
+            break;
+        }
+      }
+    }
+    self.initializationResult = static_cast<BATResult>(result);
+    for (BATBraveLedgerObserver *observer in [self.observers copy]) {
```

Comment:
Is it 100% necessary to make a copy of the observers? Seems like without it, there is a chance for race-condition.

Reply:
We do this for all observer calls as a precaution so we do not run the possibility of enumerating over `observers` while it is being mutated (this will lead to a crash).